File: parallel/ParallelEvenSum.dart
import 'dart:isolate';

var result;

int sum(List<int> a) {
  var res = 0;
  for (var i = 0; i < a.length; i++) {
    if (a[i] % 2 == 0) res += a[i];
  }
  result = res;
  return res;
}

void psum(var msg) {
  List<int> a = msg[0];
  SendPort reply = msg[1];
  int res = 0;
  for(var i = 0;i < a.length;i++){
      if(a[i] % 2 == 0)
          res += a[i];
  }
  //var res = a.where((n) => n % 2 == 0).reduce((a, b) => a + b);
  reply.send(res);
}

parallelSum(List<int> a, int N) {
  var sw = new Stopwatch()..start();
  var res = 0;
  var count = 0;
  var receivePort = new ReceivePort();
  var offset = a.length ~/ N;
  for (var i = 0; i < N; i++) {
      // Spawn does not support transferable objects presently, 
      // it has to copy list a and then transfer it to other isolates.
      // Also,sublist should also be avoided using here because sublist
      //  function will do a shallow copy from origin list.
      // Once transferable data is supported by isolate, changing to use
      //  ByteBuffer to storage data and use typed_data.view to get subdata
      //  for sending to fully avoid copying.
    Isolate.spawn(psum, [a.sublist(i * offset, i * offset + offset), receivePort.sendPort]);
  }
  receivePort.listen((msg) {
    res += msg;
    count++;
    if (count == N){
        result = res;
        sw.stop();
        print('sum using parallel: ${sw.elapsedMilliseconds}ms.');
        receivePort.close();
    }
  });
}

void measure(String text, f()) {
  var sw = new Stopwatch();
  sw.start();
  f();
  sw.stop();
  var time = sw.elapsedMilliseconds;
  print("$text: ${time}ms.");
}

main() async {
  final l = new List<int>.generate(30000000, (i) => i);
  measure('sum using loop',() => sum(l));
  measure('sum using where reduce',() => 
          l.where((n) => n % 2 == 0).reduce((a, b) => a + b));
  parallelSum(l, 3);
}

File: parallel/ParallelEvenSum.js
var Worker = require('webworker-threads').Worker;
var now = require("performance-now");

// WAITTING FOR UPDATE AND IMPLEMENTATION.
// ArrayBuffer is not accessible to a worker at present
// issue #18 https://github.com/audreyt/node-webworker-threads/issues/18 
/*var b = new Buffer(1024*1024*32);
var arrayBuffer = b.buffer.slice(b.byteOffset,b.byteOffset + b.byteLength);
var view = new Uint8Array(arrayBuffer);
for (var i = 0; i < view.length; i++) {
    view[i] = i;
}*/
var view = [];
var result;

for (var i = 0; i < 30000000; i++) {
    view.push(i);
}

measure('sum using loop',function(){sum(view);});
measure('sum using filter reduce',function(){
var res = view.filter(function(i){return i % 2 == 0;}).
    reduce(function(a,b){return a + b;},0);
    result = res;
});

var N = 3;  
var offset = view.length / N;
res = 0;
var count = 0;
start = now();
for(var i = 0;i < N;i++){
    var w =new Worker(psum);
    // slice will do a shallow copy of origin array.
    w.postMessage(view.slice(i*offset,i*offset+offset));
    // transerable:ON
    //var subBuffer = view.slice(i*offset,i*offset+offset).buffer;
    //w.postMessage(subBuffer,[subBuffer]);
    w.onmessage = function (event) {
        res += event.data;
        count++;
        if(count == N){
            elapsed = now() - start;
            result = res;
            console.log('sum using parallel: ' +  elapsed.toFixed(3) + 'ms');
        }
    };
}

function sum(array){
    var res = 0;
    for(var i = 0;i < array.length;i++){
        if(array[i] % 2 == 0) res += array[i];
    }
    result = res;
    return res;
}

function psum() {
    this.onmessage = function(e){
        /*var b = e.data;
        var view = new Uint8Array(b.buffer,b.byteOffset, b.byteLength / Uint8Array.BYTES_PER_ELEMENT);
        //console.log(view.length);*/
        var view = e.data;
        res = 0;
        for(var i = 0;i < view.length;i++){
            if (view[i] % 2 == 0) res += view[i];
        }
        postMessage(res);
        close();
    };
}

function measure(text,f){
    var start = now();
    f();
    var elapsed = now() - start;
    console.log(text + ': ' +  elapsed.toFixed(3) + 'ms');
}
